src/main/java/io/cryostat/recordings/RecordingArchiveHelper.java

```diff
@@ -1019,6 +1063,204 @@ private long getArchivedTime(String recordingName) {
+                                        tmpFile,
+                                        connectUrl,
+                                        counter + 1,
+                                        handler);
+                                return;
+                            }
+
+                            // verified no name clash at this time
+                            vertx.fileSystem()
+                                    .move(
+                                            tmpFile,
+                                            specificRecordingsPath.resolve(filename).toString(),
+                                            (res2) -> {
+                                                if (res2.failed()) {
+                                                    handler.handle(
+                                                            makeFailedAsyncResult(res2.cause()));
+                                                    return;
+                                                }
+
+                                                handler.handle(makeAsyncResult(filename));
+                                            });
+                        });
+    }
+
+    public void validateRecording(String recordingFile, Handler<AsyncResult<Void>> handler) {
+        vertx.executeBlocking(
+                event -> {
+                    try {
+                        // try loading chunk info to see if it's a valid file
+                        try (var is = new BufferedInputStream(new FileInputStream(recordingFile))) {
+                            var supplier = FlightRecordingLoader.createChunkSupplier(is);
+                            var chunks = FlightRecordingLoader.readChunkInfo(supplier);
+                            if (chunks.size() < 1) {
+                                throw new InvalidJfrFileException();
+                            }
+                        }
+                        event.complete();
+                    } catch (CouldNotLoadRecordingException | IOException e) {
+                        event.fail(e);
+                    }
+                },
+                res -> {
+                    if (res.failed()) {
+                        Throwable t;
+                        if (res.cause() instanceof CouldNotLoadRecordingException) {
+                            t =
+                                    new ApiException(
+                                            400, "Not a valid JFR recording file", res.cause());
+                        } else {
+                            t = res.cause();
+                        }
+                        vertx.fileSystem().deleteBlocking(recordingFile);
+
+                        handler.handle(makeFailedAsyncResult(t));
+                        return;
+                    }
+
+                    handler.handle(makeAsyncResult(null));
+                });
+    }
+
+    public FileUpload getTempFileUpload(Collection<FileUpload> fileUploads) {
```

Comment:
This method seems misplaced, it should live closer to the HTTP handling code I think since it's dealing specifically with file uploads and how those are encoded in a multipart form.